### PR TITLE
Update GSI analysis jobs to use COMIN/COMOUT

### PR DESCRIPTION
**Pull request by mingshichen-noaa**
# Description
 
NCO has requested that each COM variable specify whether it is an input or an output. This completes that process for the global jgdas enkf select obs job and relevant .JGLOBAL_ATMOS_ANALYSIS and JGLOBAL_ATMOS_ANALYSIS_CALC jobs.

Refs https://github.com/NOAA-EMC/global-workflow/issues/2451

# Type of change

- Maintenance (code refactor, clean-up, new CI test, etc.)

# Change characteristics
- Is this a breaking change (a change in existing functionality)? NO
- Does this change require a documentation update? NO

# How has this been tested?
 
- Clone and build on RDHPCS
- Cycled tests on Hercules
- Forecast-only tests on Hercules
 
# Checklist
- [x] Any dependent changes have been merged and published
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes
- [x] I have made corresponding changes to the documentation if necessary

### CODE REVIEW

File: jobs/JGDAS_ENKF_SELECT_OBS [hunk 1]
```diff
@@ -32,21 +32,25 @@ GPREFIX_DET="${GDUMP}.t${gcyc}z."
...
-    COM_ATMOS_HISTORY_PREV:COM_ATMOS_HISTORY_TMPL \
+    COMIN_ATMOS_ANALYSIS_PREV:COM_ATMOS_ANALYSIS_TMPL \
+    COMIN_ATMOS_HISTORY_PREV:COM_ATMOS_HISTORY_TMPL 
 
-RUN="${GDUMP}" YMD=${gPDY} HH=${gcyc} declare_from_tmpl -r COM_ATMOS_ANALYSIS_DET_PREV:COM_ATMOS_ANALYSIS_TMPL
+RUN="${GDUMP}" YMD=${gPDY} HH=${gcyc} declare_from_tmpl -r \
+	COMIN_ATMOS_ANALYSIS_DET_PREV:COM_ATMOS_ANALYSIS_TMPL
 
-mkdir -m 775 -p "${COM_ATMOS_ANALYSIS}"
+mkdir -m 775 -p "${COMOUT_ATMOS_ANALYSIS}"
```
Comment (aerorahul) [inline jobs/JGDAS_ENKF_SELECT_OBS]:
```suggestion
mkdir -p "${COMOUT_ATMOS_ANALYSIS}"
```
Just a note here and elsewhere, we don't need to specify permissions `-m 775`

**Review by aerorahul — COMMENTED** [review]
looks good. I am no longer familiar w/ the IN and OUT of EnKF related data, so would appreciate a review from @CatherineThomas-NOAA @RussTreadon-NOAA

**Review by KateFriedman-NOAA — COMMENTED** [review]
Looks good! Please resolve the shellcheck warning in `jobs/JGDAS_ENKF_SELECT_OBS` and then I can approve.

File: jobs/JGDAS_ENKF_SELECT_OBS [hunk 2]
```diff
@@ -32,21 +32,25 @@ GPREFIX_DET="${GDUMP}.t${gcyc}z."
 export GSUFFIX=".ensmean.nc"
 
 # Generate COM variables from templates
-RUN=${RUN/enkf} YMD=${PDY} HH=${cyc} declare_from_tmpl -rx COM_OBS
-MEMDIR='ensstat' YMD=${PDY} HH=${cyc} declare_from_tmpl -rx COM_ATMOS_ANALYSIS
-declare -rx COM_ATMOS_ANALYSIS_ENS="${COM_ATMOS_ANALYSIS}"
+RUN=${RUN/enkf} YMD=${PDY} HH=${cyc} declare_from_tmpl -rx \
+	COMIN_OBS:COM_OBS_TMPL
+MEMDIR='ensstat' YMD=${PDY} HH=${cyc} declare_from_tmpl -rx \
```
Comment (DavidHuber-NOAA) [inline jobs/JGDAS_ENKF_SELECT_OBS]:
```suggestion
MEMDIR='ensstat' YMD=${PDY} HH=${cyc} declare_from_tmpl -rx \
```

File: jobs/JGDAS_ENKF_SELECT_OBS [hunk 3]
```diff
@@ -32,21 +32,25 @@
...
```
[the remark this answers is not in the document]
Comment (mingshichen-noaa) [reply jobs/JGDAS_ENKF_SELECT_OBS]: @KateFriedman-NOAA 
Thanks. I have removed the trailing space.

File: jobs/JGDAS_ENKF_SELECT_OBS (hunk 1 above)
Comment (mingshichen-noaa) [reply jobs/JGDAS_ENKF_SELECT_OBS]: @aerorahul 
Thanks. I have removed "-m 755 on line 51"

File: jobs/JGDAS_ENKF_SELECT_OBS (hunk 2 above)
Comment (mingshichen-noaa) [reply jobs/JGDAS_ENKF_SELECT_OBS]: @DavidHuber-NOAA 
Thanks. I have removed the trailing space.

**Review by aerorahul — DISMISSED** [review]
looks good to me.

**Review by KateFriedman-NOAA — DISMISSED** [review]
Looks good, thanks @mingshichen-noaa !

**Review by WalterKolczynski-NOAA — DISMISSED** [review]

**Comment by aerorahul** [comment]
@mingshichen-noaa 
The tests failed overnight because of changes made in the j-job impacted the scripts called by the j-job.

**Comment by mingshichen-noaa** [comment]
@aerorahul @WalterKolczynski-NOAA @KateFriedman-NOAA 
 Since JGDAS_ENKF_SELECT_OBS job is associated with JGLOBAL_ATMOS_ANALYSIS and JGLOBAL_ATMOS_ANALYSIS_CALC jobs, I have made changing the original COM_* variables into the corresponding COMIN_* or COMOUT_* variables. Please review the adding commits.  Thanks in advance.

File: scripts/exglobal_atmos_analysis.sh [hunk 4]
```diff
@@ -92,112 +92,112 @@ export hofx_2m_sfcfile=${hofx_2m_sfcfile:-".false."}
...
-OMIBF=${OMIBF:-${COM_OBS}/${OPREFIX}omi.tm00.bufr_d${OSUFFIX}}
-MLSBF=${MLSBF:-${COM_OBS}/${OPREFIX}mls.tm00.bufr_d${OSUFFIX}}
-SMIPCP=${SMIPCP:-${COM_OBS}/${OPREFIX}spssmi.tm00.bufr_d${OSUFFIX}}
-TMIPCP=${TMIPCP:-${COM_OBS}/${OPREFIX}sptrmm.tm00.bufr_d${OSUFFIX}}
-GPSROBF=${GPSROBF:-${COM_OBS}/${OPREFIX}gpsro.tm00.bufr_d${OSUFFIX}}
-TCVITL=${TCVITL:-${COM_OBS}/${OPREFIX}syndata.tcvitals.tm00}
-B1AVHAM=${B1AVHAM:-${COM_OBS}/${OPREFIX}avcsam.tm00.bufr_d${OSUFFIX}}
-B1AVHPM=${B1AVHPM:-${COM_OBS}/${OPREFIX}avcspm.tm00.bufr_d${OSUFFIX}}
-HDOB=${HDOB:-${COM_OBS}/${OPREFIX}hdob.tm00.bufr_d${OSUFFIX}}
+PREPQC=${PREPQC:-${COMOUT_OBS}/${OPREFIX}prepbufr${OSUFFIX}}
```
Comment (KateFriedman-NOAA) [inline scripts/exglobal_atmos_analysis.sh]: Lines 95-154 should be `COMIN_OBS` and not `COMOUT_OBS`.

**Review by KateFriedman-NOAA — CHANGES_REQUESTED** [review]
Looks good for the most part. Left one comment to resolve before approving. Thanks!

File: scripts/exglobal_atmos_analysis.sh (hunk 4 above)
Comment (mingshichen-noaa) [reply scripts/exglobal_atmos_analysis.sh]: @KateFriedman-NOAA 
Thanks a lot.

**Comment by KateFriedman-NOAA** [comment]
@mingshichen-noaa Since this PR now also includes COM updates for the analysis job, please update the PR title and description to reflect the file changeset. Thanks!

**Review by KateFriedman-NOAA — DISMISSED** [review]
COM updates look good, thanks @mingshichen-noaa ! Approve pending successful completion of CI testing.

**Comment by mingshichen-noaa** [comment]
@aerorahul @KateFriedman-NOAA @AntonMFernando-NOAA 
Just now I checked  [COM to COMIN/COMOUT Tracker](https://docs.google.com/spreadsheets/d/1S4deVHNEgSR_6-q2lDOaXq4kW-4SmxurxH-q5iPVd6o/edit?pli=1&gid=0#gid=0) and found Anton planned to work on the two jobs ( JGLOBAL_ATMOS_ANALYSIS and JGLOBAL_ATMOS_ANALYSIS_CALC). Could you tell me how I should handle this case?

**Comment by KateFriedman-NOAA** [comment]
> @aerorahul @KateFriedman-NOAA @AntonMFernando-NOAA Just now I checked [COM to COMIN/COMOUT Tracker](https://docs.google.com/spreadsheets/d/1S4deVHNEgSR_6-q2lDOaXq4kW-4SmxurxH-q5iPVd6o/edit?pli=1&gid=0#gid=0) and found Anton planned to work on the two jobs ( JGLOBAL_ATMOS_ANALYSIS and JGLOBAL_ATMOS_ANALYSIS_CALC). Could you tell me how I should handle this case?

@mingshichen-noaa Thanks for checking on this! Since the change for the eobs job requires companion updates to the analysis/analysis_calc jobs at the same time then it was necessary to submit the changes together. Hopefully @AntonMFernando-NOAA hasn't already done much work on those job scripts. :)

In the future, I suggest checking the spreadsheet sooner if you think you need to update other job scripts at the same time. If someone else has already assigned themself to that job then you should reach out to them to say you can take care of it within your work or collaborate with the other developer if they have already begun making changes. Hope that helps!

**Comment by KateFriedman-NOAA** [comment]
@mingshichen-noaa please update the COM update tracking spreadsheet to reflect the analysis/analysis_calc job updates, thanks!

**Comment by mingshichen-noaa** [comment]
@KateFriedman-NOAA 
Thank you for your prompt reply and suggestion.

**Comment by mingshichen-noaa** [comment]
@KateFriedman-NOAA @AntonMFernando-NOAA 
I have updated https://docs.google.com/spreadsheets/d/1S4deVHNEgSR_6-q2lDOaXq4kW-4SmxurxH-q5iPVd6o/edit?pli=1&gid=0#gid=0 about <html>
<body>
<google-sheets-html-origin><style type="text/css"></style>
JGLOBAL_ATMOS_ANALYSIS | Mingshi | 3092 | Yes |  
-- | -- | -- | -- | --
JGLOBAL_ATMOS_ANALYSIS_CALC | Mingshi | 3092 | Yes |  


</body>
</html>

**Comment by mingshichen-noaa** [comment]
@KateFriedman-NOAA 
After I updated the title and description of the PR (#3092), I tried to update the branch. But attempt failed and got the information: "Head branch was modified. Review and try the merge again." Could you tell me how to figure it out? Thanks in advance.

**Comment by KateFriedman-NOAA** [comment]
> @KateFriedman-NOAA After I updated the title and description of the PR (#3092), I tried to update the branch. But attempt failed and got the information: "Head branch was modified. Review and try the merge again." Could you tell me how to figure it out? Thanks in advance.

Did you figure it out @mingshichen-noaa ? I see pushes from develop into your PR branch.

**Comment by mingshichen-noaa** [comment]
@KateFriedman-NOAA 
I have figured it out. Thanks.

File: ush/gaussian_sfcanl.sh [hunk 5]
```diff
@@ -133,8 +133,8 @@ export REDERR=${REDERR:-'2>'}
 ${INISCRIPT:-}
 pwd=$(pwd)
 cd "${DATA}" || exit 99
-[[ -d "${COM_ATMOS_ANALYSIS}" ]] || mkdir -p "${COM_ATMOS_ANALYSIS}"
-[[ -d "${COM_ATMOS_RESTART}" ]] || mkdir -p "${COM_ATMOS_RESTART}"
+[[ -d "${COMOUT_ATMOS_ANALYSIS}" ]] || mkdir -p "${COMOUT_ATMOS_ANALYSIS}"
+[[ -d "${COMIN_ATMOS_RESTART}" ]] || mkdir -p "${COMIN_ATMOS_RESTART}"
```
Comment (WalterKolczynski-NOAA) [inline ush/gaussian_sfcanl.sh]: Shouldn't be creating an input directory.
```suggestion
```

**Review by WalterKolczynski-NOAA — CHANGES_REQUESTED** [review]

**Comment by mingshichen-noaa** [comment]
@WalterKolczynski-NOAA 
If I change line 137 into [[ -d "${COMOUT_ATMOS_RESTART}" ]] || mkdir -p "${COMOUT_ATMOS_RESTART}",  that seems to conflict with line 80:
 80 #     input data : ${COMIN_ATMOS_RESTART}/${PDY}.${cyc}0000.sfcanl_data.tile*.nc

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA If I change line 137 into [[ -d "${COMOUT_ATMOS_RESTART}" ]] || mkdir -p "${COMOUT_ATMOS_RESTART}", that seems to conflict with line 80: 80 # input data : ${COMIN_ATMOS_RESTART}/${PDY}.${cyc}0000.sfcanl_data.tile*.nc

How so? If it is input data, the directory already exists and does not need to be created.

**Comment by WalterKolczynski-NOAA** [comment]
Changing that one line from COMIN to COMOUT clearly isn't correct:
- `COMOUT_ATMOS_RESTART` isn't defined
- You only changed the one instance, so you're making a directory that is never used
- The comments are clear about it being an input directory, so you would need to show me where it is being written *to* in order to have a COMOUT version

**Comment by mingshichen-noaa** [comment]
@WalterKolczynski-NOAA 
you are right. I should remove the line 137 because COMIN_ATMOS_RESTART is an input directory declared at line 36 of jobs/JGLOBAL_ATMOS_ANALYSIS_CALC file.

**Comment by WalterKolczynski-NOAA** [comment]
@mingshichen-noaa can you point me to your experiment directories where you tested this PR?

**Comment by mingshichen-noaa** [comment]
@WalterKolczynski-NOAA 
Today I have made some modification on usr/gaussian_sfcanl.sh. Just now I  cd sorc and ran ./build_all.sh -gu again. The recent test results are at
/work2/noaa/global/mchen/test/jobs/JGDAS_ENKF_SELECT_OBS/EXPDIR/C96C48_hybatmDA
platform is Hercules.

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA Today I have made some modification on usr/gaussian_sfcanl.sh. Just now I cd sorc and ran ./build_all.sh -gu again. The recent test results are at /work2/noaa/global/mchen/test/jobs/JGDAS_ENKF_SELECT_OBS/EXPDIR/C96C48_hybatmDA platform is Hercules.

This experiment is not progressing. Did you put `rocotorun` into your crontab?

**Comment by mingshichen-noaa** [comment]
@WalterKolczynski-NOAA 
Yesterday I made manually rocotorun and got results. Today I did crontab -e as follows:
MAILTO="mingshi.chen@noaa.gov"
*/5 * * * * /apps/contrib/rocoto/1.3.7/bin/rocotorun -d /work2/noaa/global/mchen/test/jobs/JGDAS_ENKF_SELECT_OBS/EXPDIR/C96C48_hybatmDA/C96C48_hybatmDA.db -w /work2/noaa/global/mchen/test/jobs/JGDAS_ENKF_SELECT_OBS/EXPDIR/C96C48_hybatmDA/C96C48_hybatmDA.xml

**Comment by WalterKolczynski-NOAA** [comment]
> @WalterKolczynski-NOAA Yesterday I made manually rocotorun and got results. Today I did crontab -e as follows: MAILTO="[mingshi.chen@noaa.gov](mailto:mingshi.chen@noaa.gov)" */5 * * * * /apps/contrib/rocoto/1.3.7/bin/rocotorun -d /work2/noaa/global/mchen/test/jobs/JGDAS_ENKF_SELECT_OBS/EXPDIR/C96C48_hybatmDA/C96C48_hybatmDA.db -w /work2/noaa/global/mchen/test/jobs/JGDAS_ENKF_SELECT_OBS/EXPDIR/C96C48_hybatmDA/C96C48_hybatmDA.xml

Looks like it is running now.

File: jobs/JGLOBAL_ATMOS_ANALYSIS_CALC [hunk 6]
```diff
@@ -28,16 +28,19 @@ export GPREFIX="${GDUMP}.t${gcyc}z."
 export APREFIX="${RUN}.t${cyc}z."
 export GPREFIX_ENS="${GDUMP_ENS}.t${gcyc}z."
 
-RUN=${rCDUMP} YMD=${PDY} HH=${cyc} declare_from_tmpl -rx COM_OBS
+RUN=${rCDUMP} YMD=${PDY} HH=${cyc} declare_from_tmpl -rx \
+    COMIN_OBS:COM_OBS_TMPL
 
-YMD=${PDY} HH=${cyc} declare_from_tmpl -rx COM_ATMOS_ANALYSIS COM_ATMOS_RESTART
+YMD=${PDY} HH=${cyc} declare_from_tmpl -rx
```
Comment (aerorahul) [inline jobs/JGLOBAL_ATMOS_ANALYSIS_CALC]: @mingshichen-noaa 
This line is causing the failure.  Please fix

Comment (mingshichen-noaa) [reply jobs/JGLOBAL_ATMOS_ANALYSIS_CALC]: @aerorahul 
 
After I checked codes in  jobs/JGLOBAL_ATMOS_ANALYSIS_CALC from mingshichen-noaa:feature/jgdas_enkf_select_obs_com_in_out,  the codes had been modified as follows:
 30
 31 RUN=${rCDUMP} YMD=${PDY} HH=${cyc} declare_from_tmpl -rx \
 32     COMIN_OBS:COM_OBS_TMPL
 33
 34 YMD=${PDY} HH=${cyc} declare_from_tmpl -rx
 35     COMOUT_ATMOS_ANALYSIS:COM_ATMOS_ANALYSIS_TMPL \
 36     COMIN_ATMOS_RESTART:COM_ATMOS_RESTART_TMPL
 37
 38 RUN=${GDUMP} YMD=${gPDY} HH=${gcyc} declare_from_tmpl -rx \
 39     COMIN_OBS_PREV:COM_OBS_TMPL \
 40     COMIN_ATMOS_HISTORY_PREV:COM_ATMOS_HISTORY_TMPL
 41

**Comment by mingshichen-noaa** [comment]
@aerorahul 
Do you think whether I should make a change on line 39 in jobs/JGDAS_ENKF_SELECT_OBS
from
 39 declare -rx COMOUT_ATMOS_ANALYSIS_ENS="${COMOUT_ATMOS_ANALYSIS}"
into
39 declare -rx COMOUT_ATMOS_ANALYSIS_ENS:COM_ATMOS_ANALYSIS_TMPL

File: jobs/JGLOBAL_ATMOS_ANALYSIS_CALC (hunk 6 above)
Comment (aerorahul) [inline jobs/JGLOBAL_ATMOS_ANALYSIS_CALC]:
```suggestion
YMD=${PDY} HH=${cyc} declare_from_tmpl -rx \
```
This is the fix needed.

Comment (mingshichen-noaa) [reply jobs/JGLOBAL_ATMOS_ANALYSIS_CALC]: @aerorahul 
I have changed line 34 as follows:
 34 YMD=${PDY} HH=${cyc} declare_from_tmpl -rx
 35     COMOUT_ATMOS_ANALYSIS:COM_ATMOS_ANALYSIS_TMPL \
 36     COMIN_ATMOS_RESTART:COM_ATMOS_RESTART_TMPL

Comment (mingshichen-noaa) [reply jobs/JGLOBAL_ATMOS_ANALYSIS_CALC]: @aerorahul 
My changes are located the branch: mingshichen-noaa:feature/jgdas_enkf_select_obs_com_in_out

Comment (aerorahul) [reply jobs/JGLOBAL_ATMOS_ANALYSIS_CALC]: please push them

**Comment by WalterKolczynski-NOAA** [comment]
@mingshichen-noaa Please rewind the dead jobs in your experiment so it can continue with the fix.

**Comment by mingshichen-noaa** [comment]
@aerorahul & @WalterKolczynski-NOAA ,
I have checked the testing states of C96C48_hybatmDA are SUCCEEDED. Please have a look at the testing results at 
/work2/noaa/global/mchen/test/jobs/JGDAS_ENKF_SELECT_OBS/EXPDIR/C96C48_hybatmDA

**Comment by WalterKolczynski-NOAA** [comment]
CI Tests set up to run in /lfs/h2/emc/ptmp/walter.kolczynski/PR/PR_3092/RUNTESTS on WCOSS

**Review by WalterKolczynski-NOAA — APPROVED** [review]

**Pull request merged**